File: December 2024 Leetcode Solution/December-10.py
class Solution:
    def maximumLength(self, s: str) -> int:
        def is_valid(s, n, x):
            freq = [0] * 26
            i = 0
            while i < n:
                j = i + 1
                while j < n and s[j] == s[i]:
                    j += 1
                freq[ord(s[i]) - ord('a')] += max(0, j - i - x + 1)
                if freq[ord(s[i]) - ord('a')] >= 3:
                    return True
                i = j
            return False
        
        n, l, r = len(s), 0, len(s)
        while l < r:
            mid = (l + r + 1) // 2
            if is_valid(s, n, mid):
                l = mid
            else:
                r = mid - 1
        return -1 if l == 0 else l

File: December 2024 Leetcode Solution/December-10.cpp
class Solution {
public:
    int maximumLength(string s) {
        int n = s.size(), l = 0, r = n;
        while (l < r) {
            int mid = (l + r + 1) / 2;
            if (isValid(s, n, mid)) l = mid;
            else r = mid - 1;
        }
        return l == 0 ? -1 : l;
    }

private:
    bool isValid(const string& s, int n, int x) {
        vector<int> freq(26, 0);
        for (int i = 0; i < n;) {
            int j = i + 1;
            while (j < n && s[j] == s[i]) j++;
            freq[s[i] - 'a'] += max(0, j - i - x + 1);
            if (freq[s[i] - 'a'] >= 3) return true;
            i = j;
        }
        return false;
    }
};
